homeassistant/components/freedompro/fan.py

```diff
@@ -0,0 +1,126 @@
+"""Support for Freedompro fan."""
+import json
+
+from pyfreedompro import put_state
+
+from homeassistant.components.fan import SUPPORT_SET_SPEED, FanEntity
+from homeassistant.const import CONF_API_KEY
+from homeassistant.core import callback
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Freedompro fan."""
+    api_key = entry.data[CONF_API_KEY]
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        FreedomproFan(hass, api_key, device, coordinator)
+        for device in coordinator.data
+        if device["type"] == "fan"
+    )
+
+
+class FreedomproFan(CoordinatorEntity, FanEntity):
+    """Representation of an Freedompro fan."""
+
+    def __init__(self, hass, api_key, device, coordinator):
+        """Initialize the Freedompro fan."""
+        super().__init__(coordinator)
+        self._session = aiohttp_client.async_get_clientsession(hass)
+        self._api_key = api_key
+        self._attr_name = device["name"]
+        self._attr_unique_id = device["uid"]
+        self._characteristics = device["characteristics"]
+        self._attr_device_info = {
+            "name": self.name,
+            "identifiers": {
+                (DOMAIN, self.unique_id),
+            },
+            "model": device["type"],
+            "manufacturer": "Freedompro",
+        }
+        if "rotationSpeed" in device["characteristics"]:
+            self._attr_supported_features = SUPPORT_SET_SPEED
+        self._attr_is_on = False
+        self._attr_percentage = 0
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        return self._attr_is_on
+
+    @property
+    def percentage(self):
+        """Return the current speed percentage."""
+        return self._attr_percentage
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        if "rotationSpeed" in self._characteristics:
+            return self._attr_supported_features
+        return 0
+
```

Comment:
Don't mix logic between constructor and the property function, just do:

```suggestion
    @property
    def supported_features(self):
        """Flag supported features."""
        if "rotationSpeed" in self._characteristics:
            return SUPPORT_SET_SPEED
        return 0

```